### DISCUSSION TITLE
Skyconnect

### DISCUSSION THREAD

**Discussion by pipiche38** [zigpy/zigpy#1148] [Q&A]
I have seen some communications around HA and Skyconnect.
Would it be possible to get more information on Skyconnect and zigpy ?
Is that a fully transparent integration at the zigpy library level and so the Skyconnect will behave a an elelabs or Sonoff E key ?

or is there a need for application integration 

cc: @Hedda

**Comment by Hedda** [comment]
Be sure to read and follow -> https://github.com/zigpy/zigpy/discussions/894 (maybe ask if @agners can send you guys a few hardware samples of the Home Assistant SkyConnect?).

There will be a need for application integration in the future if you want to support Silabs EFR32 adapters with "Multi-PAN RPC" firmware to work with your Zigbee plugin for Domoticz no matter if rely on zigpy or not https://github.com/zigbeefordomoticz/Domoticz-Zigbee/ 

Currently, the Home Assistant SkyConnect USB dongle does however ship with an older Silicon Labs Zigbee EmberZNet **NCP (Network Co-Processor)** based application firmware with standard EZSP interface, meaning that the current batch of Home Assistant SkyConnect USB dongles shipping today will still work out-of-the-box in the exact same way Elelabs ELR0xx USB adapters and Sonoff ZBDongle-E USB dongle, but whether or not Home Assistant SkyConnect USB dongles sold in the future will work out-of-the-box in the will depend on which type of application firmware they will then ship with by default (or the type of application firmware flashed by the end-user if an adapter is migrating or bought used), and there are already a couple of different types of alternative experimental firmware available -> https://github.com/NabuCasa/silabs-firmware/

That is mentioned on https://www.home-assistant.io/skyconnect/ but that page does not go into details just yet as this is still experimental.

Anyway, for the future, there will be at least two additional types of alternative application firmware for multi-protocol support, (both of which are [already available as experimental firmware builds today)](https://github.com/NabuCasa/silabs-firmware/), where one of those will not be backwards compatible without modifications of the Zigbee gateway software, while the other should backwards compatible with the traditional way of direct connection with existing Zigbee gateway software without modifications.

https://openthread.io/platforms/co-processor

https://www.silabs.com/documents/public/application-notes/an1333-concurrent-protocols-with-802-15-4-rcp.pdf

* The modern 'pure' **RCP (Radio Co-Processor)** option with a "Multi-PAN RPC" firmware allows concurrent multi-protocol by off-loading both the Zigbee application layer and OpenThread application layer to separate daemons/services running. This requires existing Zigbee gateway software to implement the use of Silicon Labs CPCd (CPC daemon) + Zigbeed (Zigbee daemon) daemons/services running on the host with the radio adapter and then you connect to it via a TCP socket (it is still using EZSP though), and similar for using Thread protocol you will need to implement the OpenThread daemon/service running on the host with the radio adapter. 

https://github.com/NabuCasa/silabs-firmware/tree/main/RCPMultiPAN

https://github.com/home-assistant/addons/blob/master/silabs-multiprotocol/DOCS.md#architecture

![image](https://user-images.githubusercontent.com/6320001/214878350-5b462c8d-9091-4540-b151-6341f3e6aa9f.png)

* The other alternative is a 'compromise' option with combination firmware that features both an onboard traditional Zigbee NCP stack application and an OpenThread RCP, so then you can use the Zigbee part as before, but for using the Thread protocol you will need to implement the OpenThread daemon/service running on the host.
* 
https://github.com/NabuCasa/silabs-firmware/tree/main/EmberZNetAndOpenThreadRCP/

Anyway, the specific Silicon Labs Radio Co-Processor (RCP) communication is explained here with a [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) diagram that includes OpenThread, Bluetooth, and Zigbee → [Co-Processor Communication Silicon Labs](https://docs.silabs.com/gecko-platform/4.1/service/cpc/overview)

![image](https://user-images.githubusercontent.com/6320001/214879167-7ae857bc-be5d-4ee7-a279-8a60000086c0.png)

As I understand it; when you enable multi-protocol support in the add-on inside Home Assistant it automatically flashes a Multi-PAN RCP (Radio Co-Processor) firmware to the dongle as well as enables multi-protocol support with a Silicon Labs [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) and accompanying Zigbee daemon (Zigbeed) and OpenThread Daemon (OT Daemon) that runs the Silabs EmberZNet Zigbee networking stack and OpenThread network stack respectively on the host in order to allow multi-protocol support, so in essence that new firmware is what adds a requirement of at least an additional Silicon Labs [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) and accompanying Zigbee daemon (Zigbeed) which is currently only automatically installed and available/supported if using Home Assistant OS or Home Assistant Container. 

The ZHA integration then connects to that Silicon Labs [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) over TCP/IP using a socket as that in turn has taken over the serial port connection to the Home Assistant Skyconnect USB stick, and that  Silicon Labs [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) handles the accompanying Zigbee daemon (Zigbeed) and OpenThread Daemon (OT Daemon).

If the Home Assistant Skyconnect USB stick is installed in your computer with Home Assistant OS or Home Assistant Container with enabled multiprotocol so it has flashed Multi-PAN RCP (Radio Co-Processor) firmware and you only use it for Thread protocol and not for the ZHA integration inside Home Assistant, then other implementations such as Zigbee plugin for Domoticz or Zigbee2MQTT might be able to connect remotely to that Silicon Labs Zigbee daemon (Zigbeed) over TCP/IP using a socket instead of ZHA. 

As the Zigbee plugin for Domoticz a(nd Zigbee2MQTT stand-alone) does today not install a Silicon Labs [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) and accompanying Zigbee daemon (Zigbeed) so as such you can not simply access the Home Assistant Skyconnect USB stick if it has Multi-PAN RCP (Radio Co-Processor) firmware. 

So what is missing for the Zigbee plugin for Domoticz (and Zigbee2MQTT stand-alone) to be able to work stand-alone with Home Assistant Skyconnect USB stick when it has Multi-PAN RCP (Radio Co-Processor) firmware is its own installation of Silicon Labs [CPC daemon (CPCd)](https://github.com/SiliconLabs/cpc-daemon) and accompanying Zigbee daemon (Zigbeed).

At least in theory.

PS: OpenThread page does a better job at describing the different architectural design differences between Radio Co-Processor (RCP) and Network Co-Processor (NCP) designs  -> [https://openthread.io/platforms/co-processor](https://openthread.io/platforms/co-processor)

Radio Co-Processor (RCP) 

![image](https://user-images.githubusercontent.com/6320001/214879708-d6a3d465-c3d2-4cf2-ac52-c4da179dd5b3.png)

verses Network Co-Processor (NCP) stack:

![image](https://user-images.githubusercontent.com/6320001/214879772-601fec79-7914-44d9-adfb-f1a87524b790.png)

PS: Also check out this related unofficial Home Assistant community discussion thread -> https://community.home-assistant.io/t/home-assistant-skyconnect-usb-stick-announced-will-be-compatible-with-both-zigbee-and-thread-including-matter-chip-over-thread/433594/

**Comment by puddly** [comment]
> Is that a fully transparent integration at the zigpy library level and so the Skyconnect will behave a an elelabs or Sonoff E key ?

It's just a normal Zigbee stick that ships with the usual EmberZNet firmware. Nothing new.

If you want to run the multi-PAN firmware for Thread and move the Zigbee stack onto the host, you will need to set up the multi-PAN environment with `zigbeed` and `cpcd`: https://github.com/home-assistant/addons/tree/master/silabs-multiprotocol
